### DISCUSSION TITLE
NativeLong, Pointer and PointerType

### DISCUSSION THREAD

**Discussion by DeafMan1983** [dotnet/runtime#42616] [General]
Hello everyone,

I made simple wrapper against alternative to IntPtr like I saw problem by users like this.

You can test my library for DotNet 3.1
[Visit my repository](https://github.com/DeafMan1983/Dotnet.Native.Access)!
PS: I have tested with SDL2-CS with NativeLong = works fine - completely. X11 works fine with NativeLong , too.

Enjoy and happy coding!

**Comment by Clockwork-Muse** [comment]
@DeafMan1983 - _What_ problem is this solving for users?

Note that there are proposals for a native-int type, which might cover some of that.

**Reply by tannergooding** [reply]
A native type was already implemented and will ship in C# 9 as `nint` and `nuint`.

**Reply by DeafMan1983** [reply]
But I have made class instance like `class Window : NativeLong { ... }`

@tannergooding great idea I already implemented in NativeLong ( if Statement with Windows in constructor. But why do I need nativeUint? But it is not important. I said NativeLong, Pointer and PointerType are from [JNA](https://github.com/java-native-access/jna) ( Java Native Access ) for Java coding. Java works like C#. 

But why does Java not have IntPtr or UIntPtr because it doesn't care it. C# should work like Java if you use instance of typedef structures or typedef objects should work with NativeLong.

I have made for X11 for example:
```
using System;
using System.Runtime.InteropServices;
using Native.Access;
using static TestX11.XLib;

namespace TestX11
{
    public static class XLib
    {
        public class Display : NativeLong
        {
            public Display(long value) : base(value)
            {
            }
        }

        public class Drawable : NativeLong
        {
            public Drawable(long value) : base(value)
            {
            }
        }

        public class Window : Drawable
        {
            public Window(long value) : base(value)
            {
            }
        }

        private const string LibX11 = "libX11.so.6";

        [DllImport(LibX11, EntryPoint = "XOpenDisplay")]
        private static extern long Internal_XOpenDisplay(string display_name);

        public static Display XOpenDisplay(string display_name)
        {
            return new Display(Internal_XOpenDisplay(display_name));
        }

        [DllImport(LibX11, EntryPoint = "XCloseDisplay")]
        private static extern void Internal_XCloseDisplay(long display);

        public static void XCloseDisplay(Display display)
        {
            Internal_XCloseDisplay(display.ToNative());
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            Display display = XOpenDisplay(string.Empty);

            if (display == null)
            {
                Console.WriteLine("Error: Display can't load.\n");
            }
            else
            {
                Console.WriteLine("Success: Display loads.\n");
            }

            XCloseDisplay(display);
            Console.WriteLine("Success: Display closed.");
        }
    }
}
```
Result:
![grafik](https://user-images.githubusercontent.com/57066679/94051109-5f7c5080-fdd7-11ea-857d-be3f4c5390c1.png)
No Exception with IntPtr or Structure! That is why it works with long / int like Java made with X11.

Please download my Native.Access and compile to dll and try with WinApi. If you know how do you create implements with WinApi? I really want listen about Windows Users if it throws exception then you should tell me. Because I just am trying to resolve with cross platform class "NativeLong" and it works as well anything. Sorry I have not Windows 10. I use only Ubuntu 18.04.x.

**Reply by tannergooding** [reply]
.NET doesn't need `Pointer` and `PointerType` (outside generic type parameters and reflection) because having `T*` is already legal. If the native code takes a `int*`, you can just use `int*` in C#.

As for `NativeLong`, an API has already been approved to expose as `System.Runtime.InteropServices.CLong`: <https://github.com/dotnet/runtime/issues/13788>
This will effectively be `int` on `Windows` and `nint` on `Unix`. It is only necessary for P/Invoke bindings that need to work on both Windows and Unix.

For something like Xlib which is largely Unix specific, you could just use `nint` instead and it would still be correct, it just wouldn't work on Windows (which is probably fine because the usecase on Windows is extremely limited).
